libclamav/udf.c

```diff
@@ -0,0 +1,838 @@
+            fed->icbTag.flags, fileIdentifierDescriptor);
+
+    ret = 0;
+done:
+    return ret;
+}
+
+static void dumpTag(DescriptorTag *dt)
+{
+    fprintf(stderr, "TagId = %d (0x%x)\n", dt->tagId, dt->tagId);
+    fprintf(stderr, "Version = %d (0x%x)\n", dt->descriptorVersion, dt->descriptorVersion);
+    fprintf(stderr, "Checksum = %d (0x%x)\n", dt->checksum, dt->checksum);
+    fprintf(stderr, "Serial Number = %d (0x%x)\n", dt->serialNumber, dt->serialNumber);
+
+    fprintf(stderr, "Descriptor CRC = %d (0x%x)\n", dt->descriptorCRC, dt->descriptorCRC);
+    fprintf(stderr, "Descriptor CRC Length = %d (0x%x)\n", dt->descriptorCRCLength, dt->descriptorCRCLength);
+    fprintf(stderr, "Tag Location = %d (0x%x)\n", dt->tagLocation, dt->tagLocation);
+}
+
+typedef struct {
+    uint8_t structType;
+    char standardIdentifier[5];
+    uint8_t structVersion;
+    uint8_t rest[2041];
+} GenericVolumeStructureDescriptor;
+#define NUM_GENERIC_VOLUME_DESCRIPTORS 3
+
+/*If this function fails, idx will not be updated*/
+static int skipEmptyDescriptors(cli_ctx *ctx, size_t *idxp, size_t *lastOffsetp)
+{
+    int ret         = -1;
+    uint8_t *buffer = NULL;
+    size_t idx      = *idxp;
+    bool allzeros   = true;
+    size_t i;
+
+    while (1) {
+
+        buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
+        if (NULL == buffer) {
+            goto done;
+        }
+
+        allzeros = true;
+        for (i = 0; i < VOLUME_DESCRIPTOR_SIZE; i++) {
+            if (0 != buffer[i]) {
+                allzeros = false;
+                break;
+            }
+        }
+        if (!allzeros) {
+            break;
+        }
+        idx += VOLUME_DESCRIPTOR_SIZE;
+    }
+
+    ret = 0;
+done:
+
+    *idxp        = idx;
+    *lastOffsetp = idx;
+
+    return ret;
+}
+
+typedef enum {
+    PRIMARY_VOLUME_DESCRIPTOR                   = 1,
+    IMPLEMENTATION_USE_VOLUME_DESCRIPTOR        = 4,
+    LOGICAL_VOLUME_DESCRIPTOR                   = 6,
+    PARTITION_DESCRIPTOR                        = 5,
+    UNALLOCATED_SPACE_DESCRIPTOR                = 7,
+    TERMINATING_DESCRIPTOR                      = 8,
+    LOGICAL_VOLUME_INTEGRITY_DESCRIPTOR         = 9,
+    ANCHOR_VOLUME_DESCRIPTOR_DESCRIPTOR_POINTER = 2,
+    FILE_SET_DESCRIPTOR                         = 256
+} VOLUME_DESCRIPTOR_TAG;
+
+static PrimaryVolumeDescriptor *getPrimaryVolumeDescriptor(cli_ctx *ctx, size_t *idxp, size_t *lastOffsetp)
+{
+    uint8_t *buffer               = NULL;
+    PrimaryVolumeDescriptor *test = NULL;
+    PrimaryVolumeDescriptor *ret  = NULL;
+    size_t idx                    = *idxp;
+    size_t lastOffset             = *lastOffsetp;
+
+    if (skipEmptyDescriptors(ctx, idxp, lastOffsetp)) {
+        goto done;
+    }
+
+    idx        = *idxp;
+    lastOffset = *lastOffsetp;
+
+    buffer = (uint8_t *)fmap_need_off(ctx->fmap, idx, VOLUME_DESCRIPTOR_SIZE);
```

Comment:
I'm a little concerned if this needs to be unneeded. Or maybe use `fmap_need_off_once()` instead